### ISSUE TITLE
chaos of host member in a Original dst cluster

### ISSUE THREAD

**Issue by lambdai** [envoyproxy/envoy#20953]
Background: 
We are targeting to maintain thousands of hosts in a single original_dst cluster and we are seeing the flaws of the current original_dst cluster implementation.

From a high level, an original dst cluster maintains the active upstream hosts. The host members are created upon downstream requests.

Take a closer look,
1. The main thread maintains a host_map, keyed by host string address in the shape of "10.0.0.1".
2. The worker thread sees a snapshot of the host_map in the main thread, and each worker thread maintains a connpool keyed by Host object memory address. (0x7fdeadbeef)

The host_map in mainthread and the connpool are naturally out of sync because their keys are different. 

The out-of-sync can occur when

* a worker thread can create a number of Host object before the host is posted to the mainthread and propagated to all worker threads.
* During the host GC, when the main thread decides to clean up the Host because of "not-active-requested"(note that there can be active upstream connections!), the host are erased from the host_map. The connections to this host won't be closed, not even drained.

So each worker thread may maintain an unlimited number of stale Hosts because there are connection pools within the stale hosts. Fortunately, the connpool are not leaked due to connpool is ref counted. See "envoy.reloadable_features.conn_pool_delete_when_idle".

Now that the host_map is stale anyway, I propose to maintain per worker thread independent active Host. 

Pros:
1. Kill the main thread host GC. Currently the host is evicted if two consecutive scans mark the host inactive.
2. Avoid the copy-on-write host_map at main thread. It's not ideal to copy the rest of the host when the goal is to add or remove one of the hosts.
3. A better view of the current active hosts. 
4. A predictable number of hosts: No more than ( # of concurrency) x (# of active hosts)

Cons:
1. A host with the same ip address can be duplicated in the worker threads. 

Alternative

Protect the host_map access by a mutex. This mutex needs to be held when finding or adding a host at the worker thread during serving the request.

**Comment by mattklein123** [comment]
FWIW the DNS cache code used to work like this but was moved to a mutex to avoid lots of extra memory as well as the CPU cost of doing a bunch of copies for posting. Mutex might be fine here.

**Comment by wbpcode** [comment]
> It's not ideal to copy the rest of the host when the goal is to add or remove one of the hosts.

This is to ensure all the workers can access the host map without a mutex.

I consider the host's update to be a configuration update and assume it doesn't happen frequently. And host map access directly affects data plane access and may be very frequent.

While mutex is a very good solution, not sure what the performance impact will be. Especially when the number of threads is high.

**Comment by lambdai** [comment]
Thank @mattklein123  for the feedback. It's great to know that mutex is considered as practical solution here

**Comment by lambdai** [comment]
@wbpcode A lookup only per worker host map makes generally make sense. 
However, the sync complexity is still there.

The mutex is way easier. A read write lock has the lowest cost of mental overhead. 
If the conflict is intensive, it's trivial to switch to partitioned mutex while each partition is only responsible for a small range of hashes of the addresses.

**Comment by wbpcode** [comment]
For the eds/dns clusters, we always need to sync the host from main to worker. If a per worker host map is used, the sync would not be as complicated as current implementation.
For the orgianl dst cluster, no any sync is necessary. we can maintain per worker thread independent active hosts as you proposed.
I've always been very uninformative about mutex. Although it may just be my delusion.

**Comment by lambdai** [comment]
@wbpcode Perhaps I didn't fully get your point of per `host` overhead.

Your estimtated memory print calculation, which is 32 byte-ish per host,  is based on the fact that no duplicated "HostImpl" object is created. That means various of workers don't create HostImpl independently. 

Without the sync via a mutex, I would assume each worker would create its own HostImpl (note: not shared_ptr). 

Unfortunately, The HostImpl is large. See these fields: https://github.com/envoyproxy/envoy/blob/main/source/common/upstream/upstream_impl.h#L197-L217

Could you clarify why each worker is supposed to maintain additional shared_ptr rather than additional Host object? Thanks!

**Comment by wbpcode** [comment]
> Could you clarify why each worker is supposed to maintain additional shared_ptr rather than additional Host object? Thanks!

Sorry. I didn't take the original dst cluster into count. The DNS/EDS clusters always get the lastet host lists in main and sync them to workers. So the sync operations are ok for these type clusters, right?

For the original dst clusters, yeah, may we a mutex is necessary if we cannot accept the memory comsuption of duplicate `HostImpl` objects.

**Comment by lambdai** [comment]
Yeah, original_dst is multi-producer-multi-consumer while DNS/EDS clusters are single-producer-multi-consumer.

I'd like to start with mutex impl.  The good part is that this is complelety implementation details of an original_dst cluster. If the mutex doesn't scale, it's should be too hard to switch back, or keep both.

**Comment by kyessenov** [comment]
Well, I independently came to the same conclusion. Should have read this earlier!
This is showing up as a problem for users: https://github.com/envoyproxy/envoy/issues/22218. 
I can try to write a mutex solution since that seems to be the consensus.
